frontend/src/utilities/useDraggableTable.ts

```diff
@@ -72,84 +75,101 @@ const useDraggableTable = (
-      curListItem.id === draggedItemId
-    ) {
-      return;
-    }
-    const dragId = curListItem.id;
-    const newDraggingToItemIndex = Array.from(bodyRef.current.children).findIndex(
-      (item) => item.id === dragId,
-    );
-    if (newDraggingToItemIndex !== draggingToItemIndex) {
-      const newItemOrder = moveItem([...itemOrder], draggedItemId, newDraggingToItemIndex);
-      move(newItemOrder);
-      setDraggingToItemIndex(newDraggingToItemIndex);
-      setTempItemOrder(newItemOrder);
-    }
-  };
+  const isValidDrop = React.useCallback(
+    (evt: React.DragEvent<HTMLTableSectionElement | HTMLTableRowElement>) => {
+      if (!bodyRef.current) {
+        return;
+      }
+      const ulRect = bodyRef.current.getBoundingClientRect();
+      return (
+        evt.clientX > ulRect.x &&
+        evt.clientX < ulRect.x + ulRect.width &&
+        evt.clientY > ulRect.y &&
+        evt.clientY < ulRect.y + ulRect.height
+      );
+    },
+    [],
+  );
+
+  const onDragLeave = React.useCallback<
+    UseDraggableTable['tableProps']['tbodyProps']['onDragLeave']
+  >(
+    (evt) => {
+      if (!isValidDrop(evt)) {
+        move(itemOrder);
+        setDraggingToItemIndex(-1);
+      }
+    },
+    [isValidDrop, move, itemOrder],
+  );
+
+  const onDragOver = React.useCallback<UseDraggableTable['tableProps']['tbodyProps']['onDragOver']>(
+    (evt) => {
+      evt.preventDefault();
+      if (!bodyRef.current) {
+        return;
+      }
 
-  const onDragEnd: TrProps['onDragEnd'] = (evt) => {
-    const target = evt.target as HTMLTableRowElement;
-    target.classList.remove(styles.modifiers.ghostRow);
-    target.setAttribute('aria-pressed', 'false');
-    setDraggedItemId('');
-    setDraggingToItemIndex(-1);
-    setIsDragging(false);
-  };
+      const curListItem = (evt.target as HTMLTableSectionElement).closest('tr');
+      if (
+        !curListItem ||
+        !bodyRef.current.contains(curListItem) ||
+        curListItem.id === draggedItemId
+      ) {
+        return;
+      }
+      const dragId = curListItem.id;
+      const newDraggingToItemIndex = Array.from(bodyRef.current.children).findIndex(
+        (item) => item.id === dragId,
+      );
+      if (newDraggingToItemIndex !== draggingToItemIndex) {
+        const newItemOrder = moveItem([...itemOrder], draggedItemId, newDraggingToItemIndex);
+        move(newItemOrder);
+        setDraggingToItemIndex(newDraggingToItemIndex);
+        setTempItemOrder(newItemOrder);
+      }
+    },
+    [draggedItemId, draggingToItemIndex, itemOrder, move, moveItem],
+  );
+
+  const onDrop: TrProps['onDrop'] = React.useCallback(
+    (evt: React.DragEvent<HTMLTableRowElement>) => {
+      if (isValidDrop(evt)) {
+        setItemOrder(tempItemOrder);
+      } else {
+        onDragCancel();
+      }
+    },
+    [isValidDrop, onDragCancel, setItemOrder, tempItemOrder],
+  );
+
+  const onDragEnd: TrProps['onDragEnd'] = React.useCallback(
```

Comment:
You need to be consistent with the way types are defined for the event handlers.
```suggestion
  const onDragEnd = React.useCallback<React.DragEventHandler<HTMLTableRowElement>(
```